src/Pact/Eval.hs

```diff
@@ -889,8 +1017,12 @@ guardRecursion fname m act  =
 -- | Instantiate args in body and evaluate using supplied action.
 evalUserAppBody :: Def Ref -> ([Term Name], FunType (Term Name)) -> Info -> Gas
                 -> (Term Ref -> Eval e (Term Name)) -> Eval e (Term Name)
-evalUserAppBody d@Def{..} (as',ft') ai g run =
-  guardRecursion fname (Just _dModule) $ eAdvise ai (AdviceUser (d,as')) $ dup $ appCall fa ai as' $ fmap (g,) $ run bod'
+evalUserAppBody _d@Def{..} (as',ft') ai g run =
+#ifndef NO_ADVICE
```

Comment:
I would prefer these are broken up JUST to elide advice/dup (also see comment re positive flags), e.g.
```haskell
guardRecursion fname (Just _dModule) 
#ifdef ADVICE
  $ eAdvise ai (AdviceUser (_d,as')) $ dup 
#endif
  $ appCall fa ai as' $ fmap (g,) $ run bod'